File: mobile/src/services/authService.ts
/**
 * Auth Service
 *
 * Handles all authentication-related API calls with dual-mode support:
 * - Supabase Auth (when USE_SUPABASE_AUTH is enabled)
 * - Custom JWT (legacy fallback)
 *
 * Login and password flows use Supabase SDK directly when enabled.
 * Registration always goes through the backend (student ID verification).
 *
 * @see backend/app/routers/auth.py
 * @see backend/app/schemas/auth.py
 */

import { api } from '../utils/api';
import { storage } from '../utils/storage';
import { config } from '../constants/config';
import { getSupabaseClient } from './supabase';
import type {
  User,
  VerifyStudentIdResponse,
} from '../types';

// ---------------------------------------------------------------------------
// Request / Response types that match backend Pydantic schemas exactly
// ---------------------------------------------------------------------------

/** POST /auth/login -- mirrors backend LoginRequest */
interface LoginPayload {
  identifier: string;
  password: string;
}

/**
 * POST /auth/register -- mirrors backend RegisterRequest
 * Note: `role` is not sent because only students can self-register.
 */
interface RegisterPayload {
  student_id: string;
  email: string;
  password: string;
  first_name: string;
  last_name: string;
  phone?: string;
}

/** POST /auth/verify-student-id -- mirrors backend VerifyStudentIDRequest */
interface VerifyStudentIdPayload {
  student_id: string;
  birthdate: string; // ISO 8601 date format (YYYY-MM-DD)
}

/** Backend TokenResponse shape (returned by POST /auth/login) */
interface TokenResponse {
  access_token: string;
  refresh_token?: string;
  token_type: string;
  user: User;
}

/**
 * Backend RegisterResponse shape (returned by POST /auth/register)
 *
 * In Supabase mode `tokens` is null — the user must verify their email
 * before they can sign in via Supabase SDK.
 * In legacy mode `tokens` contains access/refresh tokens.
 */
interface RegisterResponse {
  success: boolean;
  message: string;
  user: User;
  tokens: TokenResponse | null;
}

/** POST /auth/change-password -- mirrors backend PasswordChange */
interface ChangePasswordPayload {
  old_password: string;
  new_password: string;
}

/** Simple success response from change-password, logout, etc. */
interface SuccessResponse {
  success: boolean;
  message: string;
}

/** POST /auth/forgot-password -- mirrors backend ForgotPasswordRequest */
interface ForgotPasswordPayload {
  email: string;
}

/**
 * PATCH /users/{user_id} -- mirrors backend UserUpdate
 * Profile update goes through the users router, not auth.
 */
interface ProfileUpdatePayload {
  email?: string;
  first_name?: string;
  last_name?: string;
  phone?: string;
}

// ---------------------------------------------------------------------------
// Service
// ---------------------------------------------------------------------------

export const authService = {
  /**
   * Verify that a student ID exists in the university database and that
   * the provided birthdate matches the official record.
   * This is Step 1 of the student self-registration flow with 2FA verification.
   * Always uses the backend API (student records are server-side only).
   */
  async verifyStudentId(studentId: string, birthdate: string): Promise<VerifyStudentIdResponse> {
    const payload: VerifyStudentIdPayload = { student_id: studentId, birthdate };

    interface BackendVerifyResponse {
      valid: boolean;
      student_info?: {
        student_id: string;
        first_name?: string;
        last_name?: string;
        course?: string;
        year?: number;
        section?: string;
        email?: string;
        contact_number?: string;
      };
      message: string;
    }

    const response = await api.post<BackendVerifyResponse>(
      '/auth/verify-student-id',
      payload,
    );
    const backendData = response.data;

    return {
      success: backendData.valid,
      data: {
        valid: backendData.valid,
        first_name: backendData.student_info?.first_name,
        last_name: backendData.student_info?.last_name,
        course: backendData.student_info?.course,
        year: backendData.student_info?.year?.toString(),
        section: backendData.student_info?.section,
        email: backendData.student_info?.email,
        phone: backendData.student_info?.contact_number,
      },
      message: backendData.message, // Pass through backend message
    };
  },

  /**
   * Register a new student account.
   * This always goes through the backend because student ID verification
   * must happen server-side before account creation.
   *
   * In Supabase mode: backend creates Supabase Auth user + local record.
   * A verification email is sent automatically. No tokens are returned —
   * the user must verify email first, then sign in.
   *
   * In legacy mode: tokens are returned and the user is signed in immediately.
   */
  async register(data: RegisterPayload): Promise<RegisterResponse> {
    const response = await api.post<RegisterResponse>('/auth/register', data);
    const result = response.data;

    // Legacy mode: persist tokens immediately
    if (result.tokens) {
      if (result.tokens.access_token) {
        await storage.setAccessToken(result.tokens.access_token);
      }
      if (result.tokens.refresh_token) {
        await storage.setRefreshToken(result.tokens.refresh_token);
      }
    }

    if (result.user) {
      await storage.setUser(result.user);
    }

    return result;
  },

  /**
   * Authenticate with email or student ID plus password.
   *
   * Supabase mode: uses supabase.auth.signInWithPassword(), then fetches
   * the full user profile from the backend via GET /auth/me.
   *
   * Legacy mode: sends credentials to POST /auth/login and stores tokens.
   */
  async login(identifier: string, password: string): Promise<TokenResponse> {
    const supabase = getSupabaseClient();

    if (supabase && config.USE_SUPABASE_AUTH) {
      // Supabase requires an email. If identifier looks like a student ID
      // (contains a dash), resolve it to an email via the backend resolve endpoint.
      let email = identifier;
      if (identifier.includes('-') && !identifier.includes('@')) {
        try {
          const resolveResp = await api.post<{ email: string }>(
            '/auth/resolve-email',
            { identifier, password },
          );
          email = resolveResp.data.email;
        } catch (error: any) {
          // Convert resolve-email errors to user-friendly messages
          if (error.response?.status === 404) {
            throw new Error('Student ID not found. Please check your ID or register for an account.');
          }
          if (error.response?.status === 401) {
            throw new Error('Invalid Student ID or password. Please check your credentials and try again.');
          }
          if (error.response?.data?.detail) {
            throw new Error(error.response.data.detail);
          }
          // Re-throw other errors
          throw error;
        }
      }

      const { data, error } = await supabase.auth.signInWithPassword({
        email,
        password,
      });

      if (error) {
        throw new Error(error.message);
      }

      if (!data.session) {
        throw new Error('No session returned from authentication');
      }

      // Store the Supabase session tokens for the API interceptor
      await storage.setAccessToken(data.session.access_token);
      if (data.session.refresh_token) {
        await storage.setRefreshToken(data.session.refresh_token);
      }

      // Fetch the full user profile from our backend
      const user = await this.getMe();
      await storage.setUser(user);

      return {
        access_token: data.session.access_token,
        refresh_token: data.session.refresh_token,
        token_type: 'bearer',
        user,
      };
    }

    // Legacy mode
    const payload: LoginPayload = { identifier, password };
    const response = await api.post<TokenResponse>('/auth/login', payload);
    const authData = response.data;

    if (authData.access_token) {
      await storage.setAccessToken(authData.access_token);
    }
    if (authData.refresh_token) {
      await storage.setRefreshToken(authData.refresh_token);
    }
    if (authData.user) {
      await storage.setUser(authData.user);
    }

    return authData;
  },

  /**
   * Log the current user out.
   * Supabase mode: calls supabase.auth.signOut() + backend logout.
   * Legacy mode: calls backend logout only.
   * Local storage is always cleared regardless of API results.
   */
  async logout(): Promise<void> {
    try {
      const supabase = getSupabaseClient();
      if (supabase && config.USE_SUPABASE_AUTH) {
        await supabase.auth.signOut();
      }
      await api.post<SuccessResponse>('/auth/logout');
    } catch {
      // Backend/Supabase logout is best-effort
    } finally {
      await storage.clearAuth();
    }
  },

  /**
   * Retrieve the currently authenticated user's profile from the backend.
   * Checks email verification status as well.
   */
  async getMe(): Promise<User> {
    const response = await api.get<User>('/auth/me');
    return response.data;
  },

  /**
   * Refresh the access token.
   *
   * Supabase mode: uses supabase.auth.refreshSession().
   * Legacy mode: calls POST /auth/refresh.
   *
   * Note: Supabase auto-refreshes tokens via onAuthStateChange, so this
   * is mainly for manual control or legacy mode.
   */
  async refreshToken(refreshToken: string): Promise<{ access_token: string; token_type: string }> {
    const supabase = getSupabaseClient();

    if (supabase && config.USE_SUPABASE_AUTH) {
      const { data, error } = await supabase.auth.refreshSession();

      if (error) {
        throw new Error(error.message);
      }

      if (!data.session) {
        throw new Error('No session returned from refresh');
      }

      await storage.setAccessToken(data.session.access_token);
      if (data.session.refresh_token) {
        await storage.setRefreshToken(data.session.refresh_token);
      }

      return {
        access_token: data.session.access_token,
        token_type: 'bearer',
      };
    }

    // Legacy mode
    const response = await api.post<{ access_token: string; token_type: string }>(
      '/auth/refresh',
      { refresh_token: refreshToken },
    );

    const { access_token } = response.data;
    await storage.setAccessToken(access_token);

    return response.data;
  },

  /**
   * Change the current user's password.
   *
   * Supabase mode: calls supabase.auth.updateUser() + backend change-password
   * to keep both systems in sync.
   *
   * Legacy mode: calls backend change-password only.
   */
  async changePassword(oldPassword: string, newPassword: string): Promise<SuccessResponse> {
    const supabase = getSupabaseClient();

    // Always call backend (it handles Supabase sync internally)
    const payload: ChangePasswordPayload = {
      old_password: oldPassword,
      new_password: newPassword,
    };
    const response = await api.post<SuccessResponse>('/auth/change-password', payload);

    // If Supabase is enabled, also update via SDK for immediate session sync
    if (supabase && config.USE_SUPABASE_AUTH) {
      try {
        await supabase.auth.updateUser({ password: newPassword });
      } catch {
        // Backend already handled it; SDK call is best-effort
      }
    }

    return response.data;
  },

  /**
   * Request a password reset email.
   *
   * Supabase mode: uses supabase.auth.resetPasswordForEmail() which sends
   * a magic link to the user's email. The deep link redirects to the app's
   * ResetPasswordScreen where they can set a new password.
   *
   * Legacy mode: calls backend POST /auth/forgot-password.
   */
  async forgotPassword(email: string): Promise<SuccessResponse> {
    const supabase = getSupabaseClient();

    if (supabase && config.USE_SUPABASE_AUTH) {
      const { error } = await supabase.auth.resetPasswordForEmail(email, {
        redirectTo: 'iams://reset-password',
      });

      if (error) {
        throw new Error(error.message);
      }

      return {
        success: true,
        message: 'Password reset instructions sent to your email',
      };
    }

    // Legacy mode
    const payload: ForgotPasswordPayload = { email };
    const response = await api.post<SuccessResponse>('/auth/forgot-password', payload);
    return response.data;
  },

  /**
   * Complete password reset (Supabase mode only).
   * Called from ResetPasswordScreen after the user clicks the deep link.
   * The Supabase SDK automatically picks up the session from the deep link.
   */
  async resetPassword(newPassword: string): Promise<SuccessResponse> {
    const supabase = getSupabaseClient();

    if (!supabase) {
      throw new Error('Password reset requires Supabase Auth');
    }

    const { error } = await supabase.auth.updateUser({
      password: newPassword,
    });

    if (error) {
      throw new Error(error.message);
    }

    return {
      success: true,
      message: 'Password updated successfully',
    };
  },

  /**
   * Resend email verification.
   * Calls backend POST /auth/resend-verification which uses Supabase admin.
   */
  async resendVerification(email: string): Promise<SuccessResponse> {
    const response = await api.post<SuccessResponse>(
      '/auth/resend-verification',
      { email },
    );
    return response.data;
  },

  /**
   * Check the current user's email verification status.
   * Returns the user profile; caller should check user.email_verified.
   */
  async checkVerificationStatus(): Promise<User> {
    return this.getMe();
  },

  /**
   * Update the current user's profile.
   * Uses the users router (PATCH /users/{user_id}).
   */
  async updateProfile(userId: string, data: ProfileUpdatePayload): Promise<User> {
    const response = await api.patch<User>(`/users/${userId}`, data);
    const user = response.data;
    await storage.setUser(user);
    return user;
  },

  // ---------------------------------------------------------------------------
  // Local-only helpers (no network calls)
  // ---------------------------------------------------------------------------

  async getStoredUser(): Promise<User | null> {
    return storage.getUser();
  },

  async isAuthenticated(): Promise<boolean> {
    const supabase = getSupabaseClient();

    if (supabase && config.USE_SUPABASE_AUTH) {
      const { data } = await supabase.auth.getSession();
      return !!data.session;
    }

    const token = await storage.getAccessToken();
    return !!token;
  },
};

File: backend/app/services/auth_service.py
"""
Authentication Service

Business logic for authentication and authorization.
Uses local JWT authentication with bcrypt password hashing.
"""

from datetime import datetime

from sqlalchemy.orm import Session

from app.config import logger
from app.models.user import User, UserRole
from app.repositories.student_record_repository import StudentRecordRepository
from app.repositories.user_repository import UserRepository
from app.utils.exceptions import AuthenticationError, NotFoundError, ValidationError
from app.utils.security import (
    create_access_token,
    create_refresh_token,
    hash_password,
    validate_password_strength,
    verify_password,
)


class AuthService:
    """Service for authentication operations"""

    def __init__(self, db: Session):
        self.db = db
        self.user_repo = UserRepository(db)
        self.student_record_repo = StudentRecordRepository(db)

    # ------------------------------------------------------------------
    # FUN-01-01a: Check Student ID Availability
    # ------------------------------------------------------------------

    def check_student_id(self, student_id: str) -> dict:
        """
        Check if a student ID exists in university records and is available
        for registration (not already registered).

        Args:
            student_id: Student ID to check (e.g. "21-A-01234")

        Returns:
            Dictionary with exists, available, and message fields
        """
        if not student_id or len(student_id.strip()) < 3:
            return {"exists": False, "available": False, "message": "Invalid student ID format"}

        normalized = student_id.strip().upper()

        record = self.student_record_repo.get_by_student_id(normalized)
        if not record:
            return {"exists": False, "available": False, "message": "Student ID not found in university records"}

        if not record.is_active:
            return {"exists": True, "available": False, "message": "Student ID is no longer active"}

        existing_user = self.user_repo.get_by_student_id(normalized)
        if existing_user:
            return {"exists": True, "available": False, "message": "This student ID is already registered. Please login instead."}

        return {"exists": True, "available": True, "message": "Student ID found"}

    # ------------------------------------------------------------------
    # FUN-01-01: Verify Student Identity
    # ------------------------------------------------------------------

    def verify_student_id(self, student_id: str, birthdate: datetime) -> dict:
        """
        Verify student ID and birthdate against the student_records reference table.

        Two-factor verification:
          1. Student ID must exist in university database
          2. Birthdate must match the official record

        This prevents unauthorized users from viewing student information by
        entering random student IDs.

        Args:
            student_id: Student ID to verify (e.g. "21-A-012345")
            birthdate: Student's birthdate for identity verification

        Returns:
            Dictionary with validation result and student info (only if both match)
        """
        logger.info(f"Verifying student ID: {student_id}")

        if not student_id or len(student_id.strip()) < 3:
            return {"valid": False, "student_info": None, "message": "Invalid student ID format"}

        normalized = student_id.strip().upper()

        # Look up in the school's official registry
        record = self.student_record_repo.get_by_student_id(normalized)

        if not record:
            logger.warning(f"Student ID not found in registry: {normalized}")
            return {"valid": False, "student_info": None, "message": "Student ID not found in university records"}

        if not record.is_active:
            return {"valid": False, "student_info": None, "message": "Student ID is no longer active"}

        # SECURITY: Verify birthdate matches before showing student information
        if not record.birthdate:
            logger.error(f"Student record missing birthdate: {normalized}")
            return {
                "valid": False,
                "student_info": None,
                "message": "Unable to verify identity. Please contact registrar.",
            }

        # Convert datetime to date if needed
        verification_date = birthdate.date() if isinstance(birthdate, datetime) else birthdate

        if record.birthdate != verification_date:
            logger.warning(f"Birthdate mismatch for {normalized}")
            return {
                "valid": False,
                "student_info": None,
                "message": "Identity verification failed. Please check your birthdate.",
            }

        # Reject if already has an app account (prevent duplicate registration)
        existing_user = self.user_repo.get_by_student_id(normalized)
        if existing_user:
            return {
                "valid": False,
                "student_info": None,
                "message": "This student ID is already registered. Please login instead.",
            }

        # Both student ID and birthdate verified — return student info
        return {
            "valid": True,
            "student_info": {
                "student_id": record.student_id,
                "first_name": record.first_name,
                "last_name": record.last_name,
                "course": record.course,
                "year": record.year_level,
                "section": record.section,
                "email": record.email,
                "contact_number": record.contact_number,
            },
            "message": "Student ID verified successfully",
        }

    # ------------------------------------------------------------------
    # FUN-01-02: Register Student Account
    # ------------------------------------------------------------------

    def register_student(self, registration_data: dict) -> tuple[User, dict]:
        """
        Register a new student account.

        Flow:
            1. Validate student ID against school registry
            2. Validate password strength
            3. Hash password locally
            4. Create user record with password_hash
            5. Auto-enroll in matching schedules
            6. Return JWT tokens

        Args:
            registration_data: Registration data from request

        Returns:
            Tuple of (created user, response dict with tokens)
        """
        # Validate student ID exists in school registry
        # Note: Full verification (ID + birthdate) already done in Step 1 of registration flow.
        # Here we just validate that the student exists and is not already registered.
        normalized_id = registration_data["student_id"].strip().upper()

        record = self.student_record_repo.get_by_student_id(normalized_id)
        if not record:
            raise ValidationError("Student ID not found in university records")

        if not record.is_active:
            raise ValidationError("Student ID is no longer active")

        # Check if already registered (student ID or email) in a single query
        existing_user = self.user_repo.get_by_student_id_or_email(normalized_id, registration_data["email"])
        if existing_user:
            if existing_user.student_id == normalized_id:
                raise ValidationError("This student ID is already registered. Please login instead.")
            from app.utils.exceptions import DuplicateError

            raise DuplicateError("An account with this email already exists")

        # Validate password strength
        is_valid, error_msg = validate_password_strength(registration_data["password"])
        if not is_valid:
            raise ValidationError(error_msg)

        # Use official name from student_records (prevents tampering)
        user = User(
            email=registration_data["email"],
            password_hash=hash_password(registration_data["password"]),
            role=UserRole.STUDENT,
            first_name=record.first_name,
            last_name=record.last_name,
            student_id=normalized_id,
            phone=registration_data.get("phone"),
            is_active=True,
            email_verified=True,
        )
        self.db.add(user)
        self.db.flush()

        # Auto-enroll in matching schedules
        from app.services.enrollment_service import EnrollmentService

        enrollment_service = EnrollmentService(self.db)
        enrollment_service.auto_enroll_student(
            user.id,
            normalized_id,
            record=record,
            skip_duplicate_check=True,
        )

        self.db.expire_on_commit = False
        self.db.commit()
        self.db.expire_on_commit = True

        tokens = self._generate_tokens(user)

        logger.info(f"Student registered: {user.email}")
        return user, tokens

    # ------------------------------------------------------------------
    # FUN-01-03: Login
    # ------------------------------------------------------------------

    def login(self, identifier: str, password: str) -> tuple[User, dict]:
        """
        Authenticate user and generate tokens.

        Args:
            identifier: Email or student ID
            password: Plain text password

        Returns:
            Tuple of (user, tokens)
        """
        user = self.user_repo.get_by_identifier(identifier)

        if not user:
            logger.warning(f"Login failed: User not found for identifier {identifier}")
            raise AuthenticationError("Invalid email/student ID or password")

        if not user.is_active:
            logger.warning(f"Login failed: User {user.id} is inactive")
            raise AuthenticationError("User account is inactive")

        if not verify_password(password, user.password_hash):
            logger.warning(f"Login failed: Invalid password for user {user.id}")
            raise AuthenticationError("Invalid email/student ID or password")

        tokens = self._generate_tokens(user)

        logger.info(f"User logged in successfully: {user.email}")
        return user, tokens

    # ------------------------------------------------------------------
    # FUN-01-04: Refresh Token (custom JWT)
    # ------------------------------------------------------------------

    def refresh_access_token(self, refresh_token: str) -> dict:
        """
        Generate new access token from refresh token.
        """
        from app.utils.security import verify_token

        try:
            payload = verify_token(refresh_token)

            if payload.get("type") != "refresh":
                raise AuthenticationError("Invalid token type")

            user_id = payload.get("user_id")
            if not user_id:
                raise AuthenticationError("Invalid token payload")

            user = self.user_repo.get_by_id(user_id)
            if not user or not user.is_active:
                raise AuthenticationError("User not found or inactive")

            access_token = create_access_token({"user_id": str(user.id)})

            return {"access_token": access_token, "token_type": "bearer"}

        except Exception as e:
            logger.error(f"Token refresh failed: {e}")
            raise AuthenticationError("Invalid or expired refresh token") from e

    # ------------------------------------------------------------------
    # FUN-01-05: Get Current User (handled by dependency in dependencies.py)
    # ------------------------------------------------------------------

    # ------------------------------------------------------------------
    # FUN-01-06: Request Password Reset
    # ------------------------------------------------------------------

    def forgot_password(self, email: str) -> dict:
        """
        Handle forgot password request.

        Always returns a success response to prevent email enumeration.

        Args:
            email: User email address

        Returns:
            Dictionary with success message
        """
        success_msg = "If an account with that email exists, a password reset link has been sent."

        user = self.user_repo.get_by_email(email)
        if not user:
            logger.warning(f"Password reset requested for unknown email: {email}")
        else:
            logger.info(f"Password reset requested for user: {email}")

        return {
            "success": True,
            "message": success_msg,
        }

    # ------------------------------------------------------------------
    # Password & Profile management
    # ------------------------------------------------------------------

    def change_password(self, user_id: str, old_password: str, new_password: str) -> bool:
        """
        Change user password.
        """
        user = self.user_repo.get_by_id(user_id)
        if not user:
            raise NotFoundError(f"User not found: {user_id}")

        # Verify old password
        if not verify_password(old_password, user.password_hash):
            raise AuthenticationError("Current password is incorrect")

        # Validate new password
        is_valid, error_msg = validate_password_strength(new_password)
        if not is_valid:
            raise ValidationError(error_msg)

        # Update local password hash
        self.user_repo.update(user_id, {"password_hash": hash_password(new_password)})

        logger.info(f"Password changed for user: {user.email}")
        return True

    def update_profile(self, user_id: str, update_data: dict) -> User:
        """
        Update user profile (self-service, limited fields)
        """
        user = self.user_repo.get_by_id(user_id)
        if not user:
            raise NotFoundError(f"User not found: {user_id}")

        allowed_fields = {"email", "phone"}
        filtered_data = {k: v for k, v in update_data.items() if k in allowed_fields and v is not None}

        if not filtered_data:
            raise ValidationError("No valid fields to update")

        # Check for email uniqueness if email is being changed
        if "email" in filtered_data and filtered_data["email"] != user.email:
            existing = self.user_repo.get_by_email(filtered_data["email"])
            if existing:
                from app.utils.exceptions import DuplicateError

                raise DuplicateError(f"Email already in use: {filtered_data['email']}")

        updated_user = self.user_repo.update(user_id, filtered_data)
        logger.info(f"Profile updated for user: {updated_user.email}")
        return updated_user

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _generate_tokens(self, user: User) -> dict:
        """Generate access and refresh tokens (custom JWT)."""
        access_token = create_access_token({"user_id": str(user.id)})
        refresh_token = create_refresh_token({"user_id": str(user.id)})

        return {"access_token": access_token, "refresh_token": refresh_token, "token_type": "bearer"}
